src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -0,0 +1,318 @@
+        shape.append(buildAggregationShape(source.aggregations(), showFields));
+        shape.append(buildSortShape(source.sorts(), showFields));
+        return shape.toString();
+    }
+
+    /**
+     * Method to build query-section shape
+     * @param queryBuilder search request query builder
+     * @param showFields whether to append field data
+     * @return Query-section shape as String
+     */
+    static String buildQueryShape(QueryBuilder queryBuilder, Boolean showFields) {
+        if (queryBuilder == null) {
+            return EMPTY_STRING;
+        }
+        QueryShapeVisitor shapeVisitor = new QueryShapeVisitor();
+        queryBuilder.visit(shapeVisitor);
+        return shapeVisitor.prettyPrintTree(EMPTY_STRING, showFields);
+    }
+
+    /**
+     * Method to build aggregation shape
+     * @param aggregationsBuilder search request aggregation builder
+     * @param showFields whether to append field data
+     * @return Aggregation shape as String
+     */
+    static String buildAggregationShape(AggregatorFactories.Builder aggregationsBuilder, Boolean showFields) {
+        if (aggregationsBuilder == null) {
+            return EMPTY_STRING;
+        }
+        StringBuilder aggregationShape = recursiveAggregationShapeBuilder(
+            aggregationsBuilder.getAggregatorFactories(),
+            aggregationsBuilder.getPipelineAggregatorFactories(),
+            new StringBuilder(),
+            0,
+            showFields
+        );
+        return aggregationShape.toString();
+    }
+
+    static StringBuilder recursiveAggregationShapeBuilder(
+        Collection<AggregationBuilder> aggregationBuilders,
+        Collection<PipelineAggregationBuilder> pipelineAggregations,
+        StringBuilder outputBuilder,
+        int indentCount,
+        Boolean showFields
+    ) {
+        String baseIndent = ONE_SPACE_INDENT.repeat(indentCount);
+
+        //// Normal Aggregations ////
+        if (aggregationBuilders.isEmpty() == false) {
+            outputBuilder.append(baseIndent).append("aggregation:").append("\n");
+        }
+        List<String> aggShapeStrings = new ArrayList<>();
+        for (AggregationBuilder aggBuilder : aggregationBuilders) {
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.append(baseIndent).append(ONE_SPACE_INDENT).append(aggBuilder.getType());
+            if (showFields) {
+                stringBuilder.append(buildFieldDataString(AGG_FIELD_DATA_MAP.get(aggBuilder.getClass()), aggBuilder));
+            }
+            stringBuilder.append("\n");
+
+            if (aggBuilder.getSubAggregations().isEmpty() == false) {
+                // Recursive call on sub-aggregations
+                recursiveAggregationShapeBuilder(
+                    aggBuilder.getSubAggregations(),
+                    aggBuilder.getPipelineAggregations(),
+                    stringBuilder,
+                    indentCount + 2,
```

Comment:
This should be 4 with `ONE_SPACE_INDENT`

Reply:
Also, the indent might be passed as StringBuilder or Buffer by adding indent characters on top, removing them at the bottom

Reply:
Changed to StringBuilder and fixed indent size